### ISSUE TITLE
remove  'pypiwin32' from setup requirements

### ISSUE THREAD

**Issue by stonebig** [andfoy/pywinpty#52]
Mark Hammond did publish a wheel "pypiwin32" but:
- the package name (and sourceforge page) is "pywin32",
- cgohlke publish fresher wheels under "pywin32" name.

so to be able to keep cgohlke wheels, I whish you remove this line for the moment (when publishing wheels) (reverting https://github.com/spyder-ide/pywinpty/pull/49)

https://github.com/spyder-ide/pywinpty/blob/master/setup.py#L42

**Comment by andfoy** [comment]
@stonebig As long there's not an unified and official wheel release of pywin32, we can't remove this dependency. Pywinpty should work out of the box, and that includes dependencies that should be downloaded from PyPi

**Comment by stonebig** [comment]
ok, I'll look for a workaround.

**Comment by andfoy** [comment]
@ccordoba12 Should we remove pywin32 dependencies to prevent any problem with other pywin32 distributions?

**Comment by ccordoba12** [comment]
Is it possible? What are you using pywin32 for?

**Comment by goanpeca** [comment]
@andfoy for your specific use case, maybe we could use CTypes directly and not rely on pywin?

**Comment by andfoy** [comment]
That was my original intent, however, when I implement reading/writing routines directly on C, they cause several errors, that's why they were removed on #16

**Comment by stonebig** [comment]
as a matter of fact, all the sources published are apparently pointing to "pywin32" as the name of the package. so the true source of "pypiwin32" is unknown, both in:
- https://github.com/pywin32 (look for yourself)
- http://pywin32.hg.sourceforge.net/hgweb/pywin32/pywin32/summary 
- there is no account https://github.com/pypiwin32  

Example : 

 I can see in setup.py a reference to a file that is not in the published "pypiwin32" wheel: "pywin32.version.txt"
https://github.com/pywin32/pypiwin32/blob/master/setup.py#L409

(you see instead "pypiwin32.version.txt" in the wheel).

Cgohlke explained me that is why he publishes wheels as  "pywin32".

Where is the "true" source of "pypiwin32" wheels ?

**Comment by andfoy** [comment]
@ccordoba12 @goanpeca I'll try to migrate I/O functions using cytpes directly instead of using pywin32

**Comment by goanpeca** [comment]
@andfoy I mean that you can directly call the windows DLLS that handle that, instead of using PyWin32 which is a convenience library (which probably use Ctypes underneath).

**Comment by goanpeca** [comment]
Look in the code of PyWin32 for the specific methods you need and maybe you can use that logic directly.

**Comment by ghost** [comment]
The pipe name returned from `winpty_conin_name()` etc could be directly passed to builtin `open()` to open. So except some extra features, I think use `open()` is enough

**Comment by stonebig** [comment]
Twisted father confessed paternity of pypiwin32 wheels, as he need pywin32 wheels on pypi. https://twitter.com/glyph/status/896265360546029568

Indeed, it's on the setup:
https://github.com/twisted/twisted/blob/trunk/src/twisted/python/_setup.py#L109

Unless @TitanSnow trick has an easy solution, I will move from pywin32 wheel of cgohlke to pypiwin32 wheel.

**Comment by stonebig** [comment]
There has been updates to "pywin32" project, and "pypiwin32" wheels are outdated.

http://pywin32.hg.sourceforge.net/hgweb/pywin32/pywin32/raw-file/tip/CHANGES.txt

so my demand to remove the "pypiwin32" wheel dependancy (in setup.py) remains pertinent, until up-to-date wheels are available on pypi.

**Comment by ghost** [comment]
The only problem is that there seems to be no way to get length of a pipe opened by `open()`. [winpty/winpty_wrapper.py#L42](https://github.com/spyder-ide/pywinpty/blob/467a36cc1ee773ae6f401072d3b200b65411062b/winpty/winpty_wrapper.py#L42) uses `GetFileSize()` to know length of pipe which cannot be implemented with python's file object

For opening pipe and read, write, `open()` works well

A workaround is to use ctypes to call `GetFileSize()` and use python file object in other places

And, why `PTY.read()` is sometimes block and somethings non-block? It seems strange.

**Comment by andfoy** [comment]
> so my demand to remove the "pypiwin32" wheel dependancy (in setup.py) remains pertinent, until up-to-date wheels are available on pypi.

@stonebig Thanks for digging out this issue with the package developers. To prevent any package mangling, I'll migrate all functions to ``ctypes``. 

> And, why PTY.read() is sometimes block and somethings non-block? It seems strange.

As for the duality between non-blocking and blocking read that @TitanSnow mentioned, we require pywinpty callers to poll continously the output pipe to see if there are new contents available. Because ``ReadFile`` is blocking, we must first poll the pipe using ``GetFileSize`` to prevent any process block, if the length of the pipe is zero, we don't even execute ``ReadFile``, that's why is "non-blocking".

**Comment by ghost** [comment]
I think there are some better solutions on block and non-block

* `read()` does not accept a `length` argument and always read exact length of data got from `GetFileSize()`
* `read()` reads the min length of filesize and `length` arg
* `read()` always reads `length` data

The first two are non-block always, and last is block always.

**Comment by andfoy** [comment]
> ``read()`` reads the min length of filesize and length arg

I think this should be the default behaviour.

**Comment by ghost** [comment]
To apply this solution, I found that just need to modify [winpty_wrapper.py#L45](https://github.com/spyder-ide/pywinpty/blob/467a36cc1ee773ae6f401072d3b200b65411062b/winpty/winpty_wrapper.py#L45) a little

**Comment by andfoy** [comment]
Yes, I should include this as part of the cytpes migration.

**Comment by ghost** [comment]
Just in case you are wondering, I have taken over maintenance of the "pywin32" project. I would highly recommend that "pywin32" be eliminated as a dependency for as many projects as possible because it is not a safe dependency at all.

**Comment by andfoy** [comment]
@xoviat Thanks for the advice!

**Comment by ccordoba12** [comment]
@xoviat, why is not a safe dependency?

**Comment by xoviat** [comment]
The tree hosted at sourceforge is uncompilable. Think about that:
uncompilable. You can't build from source. The binaries currently hosted on
sourceforge come from a Mark Hammond's computer which has some hacks that
allow him to compile pywin32. You can also use 'wheel convert' to generate
wheels from the binaries on sourceforge.

So I spent a lot of time trying to patch the sourceforge tree so that it
would compile, which is at pywin32/pypiwin32. But that tree still has some
major problems: it uses nonstandard import machinery that has caused
problems with saltstack and other hacks that are about to be removed after
Python 2.7 dies. So then I decided to refactor the source tree into one
which is much nicer and easier to understand, which is at
pywin32/win32core, which I expect will replace the pypiwin32 package after
it comes out of beta (with an import shim).

So to sum up:
- Major portions of the sourceforge tree may stop working when python is
cleaned up after Python 2.7 dies
- The tree that I have 'cleaned up' (win32core) may be safe to use but it's
currently in beta due to lack of testing


And oh, I gave up on the PythonWin thing. I don't think anyone uses it and
it's too much to mess with at this time.

2017-08-22 13:19 GMT-05:00 Carlos Cordoba <notifications@github.com>:

**Comment by ghost** [comment]
I should mention that all of the wheels on pypiwin32 come from running
'wheel convert' on the installers that you can download from sourceforge.

2017-08-22 13:36 GMT-05:00 xoviat <xoviat@gmail.com>:
